cranelift/isle/isle/src/sema.rs

```diff
@@ -535,6 +681,161 @@ impl Expr {
+                let arg_values_tys = arg_exprs
+                    .iter()
+                    .map(|arg_expr| arg_expr.visit(visitor, termenv, vars))
+                    .zip(termdata.arg_tys.iter().copied())
+                    .collect();
+                match &termdata.kind {
+                    TermKind::EnumVariant { variant } => {
+                        visitor.add_create_variant(arg_values_tys, ty, *variant)
+                    }
+                    TermKind::Decl {
+                        constructor_kind: Some(ConstructorKind::InternalConstructor),
+                        multi,
+                        ..
+                    } => {
+                        visitor.add_construct(
+                            arg_values_tys,
+                            ty,
+                            term,
+                            /* infallible = */ false,
+                            *multi,
+                        )
+                    }
+                    TermKind::Decl {
+                        constructor_kind: Some(ConstructorKind::ExternalConstructor { .. }),
+                        pure,
+                        multi,
+                        ..
+                    } => {
+                        visitor.add_construct(
+                            arg_values_tys,
+                            ty,
+                            term,
+                            /* infallible = */ !pure,
+                            *multi,
+                        )
+                    }
+                    TermKind::Decl {
+                        constructor_kind: None,
+                        ..
+                    } => panic!("Should have been caught by typechecking"),
+                }
+            }
+        }
+    }
+
+    fn visit_in_rule<V: RuleVisitor>(
+        &self,
+        visitor: &mut V,
+        termenv: &TermEnv,
+        vars: &HashMap<VarId, V::PatternId>,
+    ) -> V::Expr {
+        let var_exprs = vars
+            .iter()
+            .map(|(&var, &val)| (var, visitor.pattern_as_expr(val)))
+            .collect();
+        visitor.add_expr(|visitor| VisitedExpr {
+            ty: self.ty(),
+            value: self.visit(visitor, termenv, &var_exprs),
+        })
+    }
+}
+
+/// Information about an expression after it has been fully visited in [RuleVisitor::add_expr].
+#[derive(Clone, Copy)]
+pub struct VisitedExpr<V: ExprVisitor> {
+    /// The type of the top-level expression.
+    pub ty: TypeId,
+    /// The identifier returned by the visitor for the top-level expression.
+    pub value: V::ExprId,
+}
+
+/// Visitor interface for [Rule]s. Visitors must be able to visit patterns by implementing
+/// [PatternVisitor], and to visit expressions by providing a type that implements [ExprVisitor].
+pub trait RuleVisitor: PatternVisitor {
+    /// The type of expression visitors constructed by [RuleVisitor::add_expr].
+    type ExprVisitor: ExprVisitor;
```

Comment:
Why have the `RuleVisitor` extend `PatternVisitor` but have a `ExprVisitor` associated type? Would it be reasonable to have a `PatternVisitor` associated type here instead?

Having both types be associated types would I think force the interface used in `lower_rule` to be a bit more consistent -- right now the expression sequence is returned while the pattern sequence is an in/out param.

Reply:
My simplest answer is "because I needed `ExprVisitor` to be a different type from `RuleVisitor` but I didn't need `PatternVisitor` to be a different type." In fact for the new IR I'm building that I want to use this in, all three types are the same, because expressions are de-duplicated across `if-let`s and even across separate rules in the same term.

That's a pretty weak justification though. There's certainly a nice symmetry argument for making both be associated types, and it could support more general usage. It also gives `RuleVisitor`s the opportunity to ignore patterns, like they currently could ignore expressions.

But making `PatternVisitor` an associated type doesn't change `lower_rule` at all, or `Rule::visit` significantly. The fundamental reason those look weird is because visiting a pattern returns nothing, while visiting an expression returns some result. I think that distinction is inherent. `PatternId`s name the inside of a pattern, top-down, so a `PatternVisitor` can record anything it might need as it goes. `ExprId`s name the outside of an expression, bottom-up, so there's a final result that the visitor might need to do something with.

It looks especially weird with the current IR because `PatternSequence::add_expr` returns ownership of an entire heap-allocated `ExprSequence`. My new IR just has to return an array index there, so I think that'll be a little less confusing.

Anyway, I've implemented this because the symmetry really is nice. It makes the `RuleVisitor` trait definition noisy with `<Self::PatternVisitor as PatternVisitor>::PatternId` but so it goes.